SignalServiceKit/src/Messages/OWSMessageSender.m

```diff
@@ -429,7 +438,11 @@ - (void)sendMessage:(TSOutgoingMessage *)message
         //
         // So we're using YDB behavior to ensure this invariant, which is a bit
         // unorthodox.
-        [message updateWithMessageState:TSOutgoingMessageStateAttemptingOut];
+        [self.dbConnection readWriteWithBlock:^(YapDatabaseReadWriteTransaction *transaction) {
+            // All outgoing messages should be saved at the time they are enqueued.
+            [message saveWithTransaction:transaction];
+            [message updateWithMessageState:TSOutgoingMessageStateAttemptingOut transaction:transaction];
```

Comment:
IIRC the "point" of `updateWithMessageState` was to avoid clobbers by applying the changes to the latest version of the saved model.

Since any potential clobbering would already be done by the previous line (though we're pretty sure the model shouldn't have been persisted until this moment**), it seems more to the point to say:

```
+  message.messageState = TSOutgoingMessageStateAttemptingOut
   [message saveWithTransaction:transaction]
-  [message updateWithMessageState:TSOutgoingMessageStateAttemptingOut transaction:transaction];
```
Is there more to it?

**On a related note, what would you think about adding an `isPersisted` attribute to TSYapModel that we could assert against?

It would get set in `initWithCoder` and in `saveWithTransaction`

Reply:
```
```
+  message.messageState = TSOutgoingMessageStateAttemptingOut
   [message saveWithTransaction:transaction]
-  [message updateWithMessageState:TSOutgoingMessageStateAttemptingOut transaction:transaction];
```
Unfortunately, this would require making the `messageState` property writable, which we don't want to do.  Making these properties `readonly` is invaluable for ensuring correct usage IMO.

```
**On a related note, what would you think about adding an isPersisted attribute to TSYapModel that we could assert against?
```
I've been thinking about that for quite a while - I think it might be quite useful for other problems.  Let's make a card.

Reply:
👍